### PR TITLE
Perf: Optimize memory usage and software large file crashes

### PR DESCRIPTION
**Pull request by zjiefee**
- Use bufferGeometry instead of geometry to reduce memory usage by 70%
- Remove toolpath object to reduce memory usage
- Remove gcode beans to reduce memory usage
- Use string array instead of string to reduce memory usage
- Use file stream read and write to reduce memory usage when file loading
- Put workspace gcode threejs object3D parsing and generation into asynchronous threads to solve the problem of software stuck
- Put laser/cnc gcode generation into server
- Add progress bar when workspace gcode loading
- Add progress bar when laser/cnc gcode generating
- Set autoPreview = false when uploadMode = greyscale

### CODE REVIEW

File: src/app/flux/cnc/index.js [hunk 1]
```diff
@@ -68,8 +70,25 @@ const INITIAL_STATE = {
...
+                    dispatch(sharedActions.onReceiveTaskResult('cnc', taskResult));
+                }
+            },
+            'taskCompleted:generateGcode': (taskResult) => {
+                if (taskResult.headerType === 'cnc') {
+                    dispatch(sharedActions.onReceiveGcodeTaskResult('cnc', taskResult));
+                }
+            },
+            'taskProgress:generateToolPath': (progress) => {
+                dispatch(sharedActions.updateState('cnc', {
```
Comment (parachvte) [inline src/app/flux/cnc/index.js]: Check head type

File: src/app/flux/cncLaserShared/index.js [hunk 2]
```diff
@@ -353,57 +368,71 @@ export const actions = {
...
+            } else {
+                return 1;
+            }
+        });
+        dispatch(actions.updateState(
+            from, {
+                isGcodeGenerating: true
+            }
+        ));
+        modelInfos[0].thumbnail = thumbnail;
```
Comment (parachvte) [inline src/app/flux/cncLaserShared/index.js]: set `thumbnail` on `orderModelInfos`.

File: src/app/flux/machine/index.js [hunk 3]
```diff
@@ -609,8 +612,9 @@ export const actions = {
             return;
         }
         const gcode = gcodeList.map(gcodeBean => gcodeBean.gcode).join('\n');
-        const filename = getGcodeName(gcodeList);
-        const type = getGcodeType(gcodeList);
+        // TODO gcodeFile
```
Comment (parachvte) [inline src/app/flux/machine/index.js]: TODO

File: src/app/widgets/LaserSetBackground/PrintSquareTrace/index.jsx [hunk 4]
```diff
@@ -106,7 +106,7 @@ class PrintSquareTrace extends PureComponent {
 
             const gcodeStr = generateSquareGcode(size, sideLength, power);
             this.props.clearGcode();
-            this.props.addGcode('Laser Coordinating G-code.nc', gcodeStr);
+            this.props.renderGcode('Laser_Coordinating_G-code.nc', gcodeStr);
```
Comment (parachvte) [inline src/app/widgets/LaserSetBackground/PrintSquareTrace/index.jsx]: Remove `G-code` ?

File: src/app/widgets/LaserTestFocus/TestFocus.jsx [hunk 5]
```diff
@@ -52,7 +52,7 @@ class TestFocus extends PureComponent {
             const jogSpeed = 1500;
             const gcode = generateLaserFocusGcode(power, workSpeed, jogSpeed);
             this.props.clearGcode();
-            this.props.addGcode('Laser Fine Tune G-code.nc', gcode);
+            this.props.renderGcode('Laser_Fine_Tune_G-code.nc', gcode);
```
Comment (parachvte) [inline src/app/widgets/LaserTestFocus/TestFocus.jsx]: `Laser_Fine_Tune.nc`

File: src/app/widgets/WifiTransport/WifiTransport.jsx [hunk 6]
```diff
@@ -4,10 +4,9 @@ import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { connect } from 'react-redux';
 import _ from 'lodash';
-import jQuery from 'jquery';
 import path from 'path';
+import jQuery from 'jquery';
```
Comment (parachvte) [inline src/app/widgets/WifiTransport/WifiTransport.jsx]: Remove jQ

File: src/app/widgets/WifiTransport/WifiTransport.jsx [hunk 7]
```diff
@@ -100,23 +98,22 @@ class WifiTransport extends PureComponent {
...
-            jQuery.get(gcodePath, (result) => {
-                this.props.clearGcode();
-                this.props.addGcode(find.name, result);
-            });
+            this.props.renderGcodeFile(find);
         },
-        onRenameDefinitionStart: (selectFileName, uploadName, index, event) => {
-            const basename = selectFileName.replace(/(.gcode|.cnc|.nc)$/, '');
-            this.props.renameGcodeFile(uploadName, basename, true);
+        onRenameDefinitionStart: (uploadName, index, event) => {
```
Comment (parachvte) [inline src/app/widgets/WifiTransport/WifiTransport.jsx]: What definition?

File: src/app/flux/cnc/index.js (hunk 1 above)
Comment (zjiefee) [reply src/app/flux/cnc/index.js]: change header type to headType

**Pull request merged**